[RFCs] Add Flow Graph try_put_and_wait RFC (#1513)
oneapi-src/oneTBB: rfcs/experimental/flow_graph_wait_for_one/README.md

{
    ComputeTool shared_tool;

    oneapi::tbb::parallel_for(0, 10,
        [&](int i)
        {
            // Preparing the input for index i

            ComputeOutput output = shared_tool.compute(input);

            // Post process output for index i
        });
}

```

The ``ComputeTool`` is a user interface that performs some computations on top of oneTBB Flow Graph. The function ``compute(input)`` should
submit the input provided to the computation graph and wait for completion of the input. 
This function also can be executed concurrently from several threads as it shown in the ``main``.

While using the ``wait_for_all()``, each thread submitting the work to the graph is required to wait
until **all** of the tasks to be done, even those that are unrelated to the input submitted. If some
post-processing is required on each thread after receiving the computation result, it would be only safe to start it when the 
Flow Graph is completed. That is inefficient since the post-processing of lightweight graph tasks are blocked by processing of unrelated inputs.

To remove this negative performance effect, the ``wait_for_one`` extension waits for completion of only one message

> (instead of the full completion of the graph):

Author: Renamed

complete.

Consider the following graph:

```cpp
using namespace oneapi::tbb;

flow::graph g;

flow::broadcast_node<int> start(g);

flow::function_node<int, int> f1(g, unlimited, f1_body);
flow::function_node<int, int> f2(g, unlimited, f2_body);

flow::join_node<std::tuple<int, int>> join(g);

flow::function_node<int, int> pf(g, serial, pf_body);

flow::make_edge(start, f1);
flow::make_edge(start, f2);

flow::make_edge(f1, flow::input_port<0>(join));
flow::make_edge(f2, flow::input_port<1>(join));

flow::make_edge(join, pf);

// Parallel submission
oneapi::tbb::parallel_for (0, 100, [](int input) {
    start.try_put(input);
});

start.try_put_and_wait(444);
// Post-processing 444

g.wait_for_all();
```

<img src="try_put_and_wait_graph.png" width=400>

Each message is broadcasted from ``start`` to two concurrent computational functions ``f1`` and ``f2``.

The result is joined into single tuple in ``join`` node and post-processed in a serial ``pf`` function node.

The task queue corresponding to each node in the graph is exposed under the node in the picture.

> Blue tasks in the queue relates to the parallel loop 0-100. Red tasks relates to the message submitted as an input to ``try_put_and_wait``.

Author: Updated both picture and the description

Tasks in `join` and `pr` are shown as partially blue and partially red since the queueing exact order of computation in ``f1`` and ``f2`` is unknown
and the queueing ``join_node`` is combining element in the order of their submission to the ports by the predecessors. Hence, it is possible that 
``join`` will create a tuple from the message that relates to ``try_put_and_wait`` (red task) and the one that does not (blue task). 

The ``try_put_and_wait`` is expected to exit when all of the red tasks and the necessary amount of blue tasks would be completed.

Completion of all blue tasks as in ``wait_for_all`` is not guaranteed.

## Feature Design

This feature is implemented by creating an instance of special private class ``message_metainfo`` for each input message in ``try_put_and_wait``. This instance

> wraps the pointer to the dedicated ``wait_context`` object representing the reference counter for single input message.

Reviewer: I did not find any description of how the reference counter is used - that is, at which point(s) it is incremented and decremented.

metainformation instance is buffered as well. While buffering the metainfo, the additional reference counter is held on the underlying ``wait_context`` to make sure ``try_put_and_wait`` would
block until the message would be retrieved from the buffer. 

``message_metainfo`` class may be reused in the future to support additional use-cases when it is required to push additional information about the input message through the graph. E.g. supporting
priorities for single messages - the corresponding priority tag can be assigned as part of the metainfo.

Metainformation class supports containing multiple reference counters at the same time to support joining multiple messages with different associated reference counters in the ``join_node``. 
See [separate section](#details-about-metainformation-class) for more details.

Implementation-wise, processing the metainformation is exposed by adding new internal virtual member functions to various Flow Graph instance:

``` cpp
template <typename T>
class receiver {
protected:
    virtual bool try_put_task(const T& t) = 0; // Existing API

>     virtual bool try_put_task(const T& t, const message_metainfo& info); // New API

Author: Fixed

``wait_context``s stored in ``metainfo`` and also wraps the ``metainfo`` object itself since it would be broadcasted to the successors when the task is completed.

If the concurrency is not ``unlimited``, the call to ``try_put_task`` tries to occupy the concurrency of the node. If the thread limit is not yet reached - behaves the same as 
in the ``unlimited`` case. Otherwise, both input message and the metainfo are stored in the internal queue of the node. When one of the node tasks is completed, it retrieves 
the postponed message and the corresponding metainfo from the queue and spawns a task to process it.

Since the ``function_node`` guarantees that all of the elements would be retrieved from the internal queue at some time, [buffering issues](#buffering-the-metainfo) cannot take place.

### Rejecting ``function_node``

If the concurrency of the node is set to ``unlimited``, behaves the same as in the ``queueing`` case described above.

Otherwise, if the concurrency limit of the node is reached, both message and the associated metainformations would be rejected and the predecessor that called the ``try_put_task``
is responsible on buffering both of them.

> If the predecessor is not the buffering node, both message and the metainfo would be lost.

Reviewer 2: You should say what "lost" means in this context. How does it affect waiting.

Author: Updated the description

Since there are no tasks, the calling thread would broadcast the output and the metainformation to the successors after completing the function.

### ``continue_node``

``continue_node`` only executes the associated body (and broadcasts the signal to the successors) when it receives ``N`` signals from the predecessors, where ``N``
is the number of predecessors. 

It means that prior to executing the body, the node can receive several ``metainfo`` instances from different predecessors. To handle this, the node initially stores an
empty metainfo instance inside itself and each call to ``try_put_task`` with non-empty metainfo, merges the received metainformation with the stored instance.

Additional reference counter would be held on each input ``wait_context`` to make sure the corresponding ``try_put_and_wait`` will remain blocked until the item would leave the
``continue_node``. 

When the ``continue_node`` receives the last signal from the predecessors, it creates a task to complete the associated body. Once this task is executed, the metainfo instance stored in

> the ``continue_node`` switches back to the empty state for further work and the previously accumulated metainfo is broadcasted to the successors of the node.

Reviewer: As described, there seems to be a race in case the `continue_node` can be triggered more than once during an execution of the graph. In such case, new messages might come while the body of the node is executed - so switching the state back to empty should seemingly happen before executing the body.

Author: Thanks, it is an incorrect description. The implementation resets the stored metainfo before spawning the task under the node's mutex. 
Updated the description.

both of them are removed from the buffer. Otherwise, the push-pull protocol works and the successor should return for the item once it becomes available by calling
``try_get(msg, metainfo)`` or ``try_reserve(msg, metainfo)``.

Since placing the buffers before rejecting nodes is not the only use-case, there is a risk of issues relates to buffering. It is described in details in the [separate section](#buffering-the-metainfo).

### Broadcast-push buffering nodes

The issue with broadcast-push ``overwrite_node`` and ``write_once_node`` is these nodes stores the received item and even if this item is accepted by one of the successors,
it would be broadcasted to others and kept in the buffer.

Since the metainformation is kept in the buffer together with the message itself, even if the computation is completed, the ``try_put_and_wait`` would stuck because of the reference
held by the buffer.

Even the ``wait_for_all()`` call would be able to finish in this case since it counting only the tasks in progress and ``try_put_and_wait`` would still be blocked.

``try_put_and_wait`` feature for the graph containing these nodes should be used carefully because of this issue:

> * The ``overwrite_node`` should be explicitly reset by calling ``node.reset()`` or the element with the stored metainfo should be overwritten with another element.
> * The ``write_once_node`` should be explicitly reset by calling ``node.reset()`` since the item cannot be overwritten.

Reviewer 2: ```suggestion
* The ``overwrite_node`` should be explicitly reset by calling ``node.clear()`` or the element with the stored metainfo should be overwritten with another element.
* The ``write_once_node`` should be explicitly reset by calling ``node.clear()`` since the item cannot be overwritten.
```

### ``broadcast_node``

While ``broadcast_node::try_put_task`` is called with the metainfo argument - both item and the associated metainformation would be broadcasted to each successor of the node.

### ``limiter_node``

If the threshold of the node was not reached, both value and the metainformation should be broadcasted to the successors. Otherwise- both should be rejected and buffered by another node. 

Metainformation on the decrement port is ignored since this signal should not be considered part of working on the original message.

### Queueing ``join_node``

> Each input port of the join_node should support the queue for both values and the associated metainformations. Once all of the input ports would contain the value, the values

Reviewer 2: ```suggestion
Each input port of the join_node should support the queue for both values and the associated metainformations. Once all of the input ports contain the value, the values
```

should be combined into single tuple output and the metainformation objects should be combined into single metainfo using `metainfo1.merge(metainfo2)`, associated with the tuple
and submitted to successors.

> If the item with the metainformation is stored in the internal queue of one of the input ports, but items never received by other ports, the item and the metainformation would be kept in the
> queue and block the corresponding ``try_put_and_wait`` call.

Reviewer 2: ```suggestion
If an item with the metainformation is stored in the internal queue of one of the input ports, but items are never received by other ports, the item and the metainformation will be kept in the
queue and block the corresponding ``try_put_and_wait`` call.
```

Reviewer 2: We may need to add a `clear` function to `join_node` to allow explicit clearing of the internal queues.

```cpp
using namespace oneapi::tbb;

flow::function_node<int, int> start1(g, ...);
flow::function_node<int, int> start2(g, ...);

flow::join_node<std::tuple<int, int>> join(g);

flow::function_node<std::tuple<int, int>, int> post_process(g, ...);

flow::make_edge(start, flow::input_port<0>(join));
flow::make_edge(start, flow::input_port<1>(join));
flow::make_edge(join, post_process);

std::thread t1([&]() {
    start1.try_put_and_wait(1);
});

std::thread t2([&]() {
    start2.try_put_and_wait(2);
})
```

### Combined or separated wait

Current proposal describes only the case where submitting the work into the flow graph and waiting for it are combined in a single use API `node.try_put_and_wait`.
In theory, it can be useful to also have the ability to split these phases:

```cpp

oneapi::tbb::flow::graph g;
oneapi::tbb::flow::function_node<int, int> start_node(g, ...);

auto desc = node.try_put(work); // other API name can be selected

// Other work
// May submit more work into the Flow Graph
// May create other descriptors that would be waited later

g.wait(desc);

```

> In that case it would be needed to extend the node with the new API returning some descriptor that can be used as a hit work waiting function `g.wait` that also should be added.

Reviewer: ```suggestion
In that case it would be needed to extend the node with the new API returning some descriptor that can be used as the argument to the work waiting function `g.wait` that also should be added.
```

An other interesting use-case is when the `multifunction_node` is used as a reduction for multiple input values with only one output. In that case the metainformation
should not be automatically propagated at all and should be accumulated simultaneously with the outputs.

Current implementation does not support ``multifunction_node`` and ``async_node``. 

Possible approaches for implementing such support:

* Hiding metainfo inside of ``output_ports`` to preserve automatic metainfo propagation.
* Merging the input and the ``message_metainfo`` together into some publicly available type ``tagged_input<T>`` and require the user to explicitly specify this type
for multi-output nodes.
* Introducing an extra unspecified type ``node_type::tag_type`` and require the user to accept it as a third argument of the body. 

## Open Questions

The following questions should be addressed before moving this feature to ``supported``:

* Multi-output nodes support should be described and implemented
* Feedback from the customers should be received
* More multithreaded tests should be implemented for the existing functionality

> * The corresponding oneAPI specification update should be done

Reviewer 2: Should we add a `clear` function to all buffering nodes?

Author: Added to the list of the questions, together with the safety guarantees for this method.

# Waiting for single messages in the Flow Graph

> Extending the oneTBB Flow Graph interface with the new ``try_put_and_wait(msg)`` API which allows to wait
> for completion of the chain of tasks corresponding to the ``msg``.

Reviewer 2: ```suggestion
This feature extends the oneTBB Flow Graph interface with a ``try_put_and_wait(msg)`` API that supports waiting
for completion of the chain of tasks related to ``msg``.
```

> The feature should improve the Flow Graph performance in scenarios where multiple threads submits the work into the
> Flow Graph simultaneously and each of them should

Reviewer 2: ```suggestion
The feature may improve Flow Graph performance in scenarios where multiple threads submit work into the same
Flow Graph and each of them need to wait for only the work associated with their message to complete.
```

## Introduction

> Current API of oneTBB Flow Graph allows doing two basic actions after building the graph:

Reviewer 2: ```suggestion
Without this feature, the oneTBB Flow Graph supports two basic actions after building the graph:
```

- Submitting messages in some nodes using the ``node.try_put(msg)`` API.

> - Waiting for completion of **all** messages in ghe graph using the ``graph.wait_for_all()`` API.

Reviewer 2: ```suggestion
- Waiting for completion of **all** messages in the graph using the ``graph.wait_for_all()`` API.
```

Since the only API currently available for waiting until the work would be completed waits for all tasks in the graph
to complete, there can be negative performance impact in use cases when the thread submitting the work should be notified as soon as possible

> when the work is done. Having only ``graph.wait_for_all()`` forces the thread to wait until **all** of the tasks in the Flow Graph, no meter
> corresponds to the waited message or not.
> 

Reviewer 2: ```suggestion
when the work is done. Having only ``graph.wait_for_all()`` forces the thread to wait until **all** of the tasks in the Flow Graph complete,
even those that are unrelated to the message it submitted.

```

    // Other Flow Graph nodes
public:
    ComputeTool()
    {
        // Builds the Flow Graph
    }

    // Performs the computation using the user-provided input
    ComputeOutput compute(ComputeInput input)
    {
        m_start_node.try_put(input); // Submit work in the graph
        m_graph.wait_for_all(); // Waiting for input to be processed
    }
};

int main()
{
    ComputeTool shared_tool;

    oneapi::tbb::parallel_for(0, 10,
        [&](int i)
        {
            // Preparing the input for index i

            ComputeOutput output = shared_tool.compute(input);

            // Post process output for index i
        });
}

```

The ``ComputeTool`` is a user interface that performs some computations on top of oneTBB Flow Graph. The function ``compute(input)`` should
perform the computations for the provided input. Since the Flow Graph is used, it should submit the message into the graph and wait for its
completion. This function also can be called concurrently from several threads as it shown in the ``main``.

Since the only available API in the Flow Graph is ``wait_for_all()``, each thread submitting the work to the graph would be required to wait

> until **all** of the tasks would be done, no meter if these tasks corresponds to the input processed by this thread or not. If some

Reviewer 2: ```suggestion
until **all** of the tasks are done, no matter if these tasks correspond to the input processed by this thread or not. If some
```

> post-processing is required on each thread after receiving the computation result, it would be only possible to start it when the Flow Graph would be completed what can be inefficient if the post-processing of lightweight graph tasks would be blocked by processing the more mature input.
> 

Reviewer 2: ```suggestion
post-processing is required on each thread after receiving the computation result, it would be only safe to start it when the Flow Graph is complete. 
This is inefficient since the post-processing of lightweight graph tasks are blocked by processing of unrelated inputs.

```

> To get rid of this negative performance effect, it would be useful to add some kind of new API to the Flow Graph that would wait for
> completion of only one message (instead of the full completion of the graph):

Reviewer 2: ```suggestion
To remove this negative performance effect, the ``wait_for_one`` extension waits for
completion of only one message (instead of the full completion of the graph):
```

```cpp
ComputeOutput compute(ComputeInput input)
{
    m_start_node.try_put(input); // Submit work in the graph
    g.wait_for_one(input); // Pseudo-code. Wait for completion of only messages for computations of input
}
```

> ## Proposal

Reviewer 2: ```suggestion
## Overview of experimental feature
```

> The idea of this proposal is to extend the existing Flow Graph API with the new member function of each receiver nodes -

Reviewer 2: ```suggestion
This feature extends the existing Flow Graph API with an additional member function for each of the receiver nodes -
```

> ``node.try_put_and_wait(msg)``. This function should submit the msg into the Flow Graph (similarly to ``try_put()``) and wait for its completion. The function should be exited only if all of the tasks corresponds to the ``msg`` and skip waiting for any other tasks to
> complete.
> 

Reviewer 2: ```suggestion
``node.try_put_and_wait(msg)``. This function submits the msg into the Flow Graph (similarly to ``try_put()``) and wait for its completion. 
The function blocks until all of the tasks related to processing ``msg`` complete and is allowed to skip waiting for any 
other tasks to complete.

```

Consider the following graph:

```cpp

using namespace oneapi::tbb;

flow::graph g;

flow::broadcast_node<int> start(g);

flow::function_node<int, int> f1(g, unlimited, f1_body);
flow::function_node<int, int> f2(g, unlimited, f2_body);
flow::function_node<int, int> f3(g, unlimited, f3_body);

flow::join_node<std::tuple<int, int, int>> join(g);

flow::function_node<int, int> pf(g, serial, pf_body);

flow::make_edge(start, f1);
flow::make_edge(start, f2);
flow::make_edge(start, f3);

flow::make_edge(f1, flow::input_port<0>(join));
flow::make_edge(f2, flow::input_port<1>(join));
flow::make_edge(f3, flow::input_port<2>(join));

flow::make_edge(join, pf);

// Parallel submission
oneapi::tbb::parallel_for (0, 100, [](int input) {
    start.try_put(input);
});

start.try_put_and_wait(444);
// Post-processing 444

g.wait_for_all();
```

> Each message is broadcasted from ``start`` to three concurrent computational functions ``f1``, ``f2`` and ``f3``. The result is when joined into single tuple in ``join`` node and
> post-processed in a serial ``pf`` function node. The task queue corresponding to each node in the graph is exposed under the node in the picture. The tasks that corresponds
> to the parallel loop 0-100 are shown as blue tasks in the queue. Red tasks corresponds to the message submitted as an argument in ``try_put_and_wait``.
> The ``try_put_and_wait`` is expected to exit when all of the red tasks and the necessary amount of blue tasks would be completed. Completion of all blue tasks as in ``wait_for_all``
> is not guaranteed.

Reviewer 2: ```suggestion
Each message is broadcasted from ``start`` to three concurrent computational functions ``f1``, ``f2`` and ``f3``. 
The result is joined into single tuple in ``join`` node and post-processed in a serial ``pf`` function node. 
The task queue corresponding to each node in the graph is exposed under the node in the picture. 
The tasks that corresponds to the parallel loop 0-100 are shown as blue tasks in the queue. Red tasks 
correspond to the message submitted as an argument in ``try_put_and_wait``. The ``try_put_and_wait`` is 
expected to exit when all of the red tasks and the necessary amount of blue tasks are completed. Completion 
of all blue tasks as in ``wait_for_all`` is not guaranteed.
```

> From the implementation perspective, the feature is implemented currently by creating an instance of special class ``message_metainfo`` with the input message in ``try_put_and_wait``

Reviewer 2: ```suggestion
## Feature Design

From the implementation perspective, the feature is implemented currently by creating an instance of special class ``message_metainfo`` with the input message in ``try_put_and_wait``
```
